### PR TITLE
documentation: add android-channel-builder.md

### PR DESCRIPTION
**Pull request by ericgribkoff**
Addresses https://github.com/grpc/grpc-java/issues/4011.

cc @makdharma @srini100

### CODE REVIEW

**Comment by srini100** [comment]
A future enhancement would be to provide an API to set the network state in gRPC so that gRPC can automatically take care of resetting the exponential backoff and few other things like cancelling existing RPCs and immediately returning error on new RPCs. If such an API already exists then it should be included in this document.

**Comment by srini100** [comment]
Is there a future plan for OkHttp to handle network events similar to Cronet? If not, is there a plan to let gRPC listen to ConnectivityManager and do some of these things transparently?

File: documentation/android-channel-builder.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+Since gRPC's 1.12 release, the `grpc-android` package provides access to the
+`AndroidChannelBuilder` class. Given an Android Context, this builder will
+register a network event listener upon channel construction.  The listener is
+used to automatically respond to changes in the device's network state, avoiding
+delays and interrupted RPCs that may otherwise occur.
+
+By default, gRPC uses exponential backoff to recover from connection failures.
+Depending on the scheduled backoff delay when the device regains connectivity,
+this could result in a  one minute or longer delay before gRPC re-establishes
+the connection. This delay is removed the `AndroidChannelBuilder` is provided
```
Comment (srini100) [inline documentation/android-channel-builder.md]: the --> when

File: documentation/android-channel-builder.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+(typically 30 second) interval when both cellular and wifi networks remain
+available, then any connections on the cellular network are terminated.  By
+listening for changes in the device's default network, `AndroidChannelBuilder`
+sends new RPCs via wifi rather than using an already-established cellular
+connection. Without listening for pending network changes, new RPCs sent on an
+already established cellular connection would fail when the device terminates
+cellular connections.
+
+***Note:*** *Currently, `AndroidChannelBuilder` is only compatible with gRPC
+OkHttp. We plan to support the gRPC Cronet transport in the future, but the
```
Comment (srini100) [inline documentation/android-channel-builder.md]: If this is only necessary with OkHTTP then what support does Cronet need?

File: documentation/android-network-management.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+two API methods to allow the application to relay information about network
+changes (e.g., monitored via Android's
+[`ConnectivityManager`](https://developer.android.com/reference/android/net/ConnectivityManager.html))
+to the gRPC channel.
+
+***Note:*** *The information and API methods discussed in this document only
+apply to the gRPC OkHttp transport. The gRPC Cronet transport (using Chromium's
+networking stack) handles connection management and monitors the device network
+state directly, outside of gRPC library code.*
+
```
[the remark this answers is not in the document]
Comment (ericgribkoff) [reply documentation/android-network-management.md]: Not sure if these comments were saved from an earlier review or not: the android-network-management.md doc is no longer included in this PR, as it's not necessary for end-users to understand how and when to call `enterIdle` and `resetConnectBackoff`. These instructions are superseded by the documentation for AndroidChannelBuilder.

[the remark this answers is not in the document]
Comment (srini100) [reply documentation/android-network-management.md]: Ignore. These are comments for earlier version. Not sure why it is showing up now.

**Pull request merged**